dll/ntdll/rtl/libsupp.c

```diff
@@ -990,6 +993,180 @@ RtlDosApplyFileIsolationRedirection_Ustr(IN ULONG Flags,
+    {
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    if (!NewName)
+    {
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    PCUNICODE_STRING PrevNewName = *NewName;
+
+
+    *RedirectedDll = FALSE;
+
+    UNICODE_STRING ApisetName = {0};
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    DWORD ApisetVersion = LdrpApisetVersion();
+    if (ApisetVersion)
+    {
+        Status = ApiSetResolveToHost(ApisetVersion, OriginalName, RedirectedDll, &ApisetName);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("ApiSetResolveToHost FAILED: (Status 0x%x)\n", Status);
+            return Status;
+        }
+    }
+
+    if (*RedirectedDll)
+    {
+        UNICODE_STRING NtSystemRoot;
+        UNICODE_STRING System32 = RTL_CONSTANT_STRING(L"\\System32\\");
+        PUNICODE_STRING ResultPath = NULL;
+
+        /* This is an apiset we can use */
+        RtlInitUnicodeString(&NtSystemRoot, SharedUserData->NtSystemRoot);
+
+        SIZE_T Needed = System32.Length + ApisetName.Length + NtSystemRoot.Length + sizeof(UNICODE_NULL);
+
+        if (StaticString && StaticString->MaximumLength >= (USHORT)Needed)
+        {
+            StaticString->Length = 0;
+            ResultPath = StaticString;
+        }
+        else if (DynamicString)
+        {
+            DynamicString->Buffer = RtlpAllocateStringMemory(Needed, TAG_USTR);
+            if (DynamicString->Buffer == NULL)
+            {
+                DPRINT1("LdrpApplyFileNameRedirection out of memory\n");
+                return STATUS_NO_MEMORY;
+            }
+            DynamicString->MaximumLength = (USHORT)Needed;
+            DynamicString->Length = 0;
+
+            ResultPath = DynamicString;
+        }
+        else
+        {
+            DPRINT1("ERROR: LdrpApplyFileNameRedirection no inputbuffer valid\n");
+            return STATUS_INVALID_PARAMETER;
+        }
+
+        RtlAppendUnicodeStringToString(ResultPath, &NtSystemRoot);
+        RtlAppendUnicodeStringToString(ResultPath, &System32);
+        RtlAppendUnicodeStringToString(ResultPath, &ApisetName);
+        DPRINT1("ApiSetResolveToHost redirected %wZ to %wZ\n", OriginalName, ResultPath);
+        *NewName = ResultPath;
+    }
+    else
+    {
+        /* Check if the SxS Assemblies specify another file */
+        Status = RtlDosApplyFileIsolationRedirection_Ustr(
+            RTL_DOS_APPLY_FILE_REDIRECTION_USTR_FLAG_RESPECT_DOT_LOCAL, OriginalName, Extension, StaticString, DynamicString, NewName, NULL, NULL, NULL);
+
+        /* Check success */
+        if (NT_SUCCESS(Status))
+        {
+            /* Let Ldrp know */
+            *RedirectedDll = TRUE;
+        }
+        else if (Status != STATUS_SXS_KEY_NOT_FOUND)
+        {
+            /* Unrecoverable SxS failure; did we get a string? */
+            if (DynamicString && DynamicString->Buffer)
+                RtlFreeUnicodeString(DynamicString);
+            return Status;
+        }
+        else /* Status == STATUS_SXS_KEY_NOT_FOUND */
```

Comment:
Moving this one up / checking for Status == STATUS_SXS_KEY_NOT_FOUND seems more intuitive to me.